File: TinySocket/Classes/socket.swift
//
//  socket.swift
//  TinySocket
//
//  Created by hao yin on 2021/1/22.
//

import Foundation

@_silgen_name("tiny_tcp")
public func tiny_tcp(domain:Int)->Int

@_silgen_name("tiny_udp")
public func tiny_udp(domain:Int)->Int

@_silgen_name("tiny_host_net_family")
public func tiny_host_net_family(family:Int)->Int

@_silgen_name("tiny_tcp_connect")
public func tiny_tcp_connect(tcp:Int,domain:Int,ip:UnsafePointer<UInt8>,port:UInt16)->Int

@_silgen_name("tiny_send")
public func tiny_send(socket:Int,data:UnsafePointer<UInt8>,size:Int)->Int

@_silgen_name("tiny_send_string_to")
public func tiny_send_string_to(socket:Int,
                                domain:Int,
                                ip:UnsafePointer<UInt8>,
                                port:UInt16,
                                char:UnsafePointer<UInt8>)

@_silgen_name("tiny_send_to")
public func tiny_send_to(socket:Int,
                        domain:Int,
                        ip:UnsafePointer<UInt8>,
                        port:UInt16,
                        data:UnsafePointer<UInt8>,
                        size:Int)->Int

@_silgen_name("tiny_recv")
public func tiny_recv(socket:Int,data:UnsafePointer<UInt8>,size:Int)->Int

@_silgen_name("tiny_recv_from")
public func tiny_recv_from(socket:Int,
                           data:UnsafePointer<UInt8>,
                           size:Int,
                           client:UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>,
                           len:UnsafeMutablePointer<Int>)->Int

@_silgen_name("tiny_send_string")
public func tiny_send_string(socket:Int,data:UnsafePointer<UInt8>)->Int

@_silgen_name("tiny_close")
public func tiny_close(socket:Int)->Int

@_silgen_name("tiny_tcp_accept")
public func tiny_tcp_accept(tcp:Int,
                            client:UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>,
                            len:UnsafeMutablePointer<UInt>)->Int

@_silgen_name("tiny_tcp_listen")
public func tiny_tcp_listen(tcp:Int,count:Int)->Int

@_silgen_name("tiny_tcp_bind")
public func tiny_tcp_bind(tcp:Int,domain:Int,port:UInt16)->Int

@_silgen_name("tiny_addr_famaly")
public func tiny_addr_famaly(addr:UnsafePointer<UInt8>)->Int

@_silgen_name("tiny_addr_port")
public func tiny_addr_port(addr:UnsafePointer<UInt8>,size:size_t)->UInt16


@_silgen_name("tiny_addr_ip")
public func tiny_addr_ip(addr:UnsafePointer<UInt8>,size:Int)->UnsafePointer<UInt8>

@_silgen_name("tiny_addr_create")
public func tiny_addr_create(domain:Int ,addr:UnsafePointer<UInt8>,port:ushort)->UnsafePointer<UInt8>


@_silgen_name("tiny_send_timeout")
public func tiny_send_timeout(tcp:Int,seconds:Int)

@_silgen_name("tiny_connect_timeout")
public func tiny_connect_timeout(tcp:Int,seconds:Int)

@_silgen_name("tiny_recv_timeout")
public func tiny_recv_timeout(tcp:Int,seconds:Int)


public enum SocketDomain{
    case SocketIpv4
    case SocketIpv6
}

public struct SocketError:Error{
    public var code:Int
    public var msg:String
    public init(code:Int,msg:String){
        self.code = code
        self.msg = msg
    }
}
public struct SocketHost{
    public var name:String
    public var aliases:[String]
    public var length:Int
    public var addrType:SocketDomain
    public var ipStrAddr:[String]
    public var ipAddr:[Data]
}

public struct SocketAddress{
    public var origin:Data
    public var domain:SocketDomain?{
        let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: origin.count)
        self.origin.copyBytes(to: ptr, count: origin.count)
        let s = tiny_addr_famaly(addr: ptr)
        if(s == AF_INET){
            ptr.deallocate()
            return .SocketIpv4
        }else if(s == AF_INET6){
            ptr.deallocate()
            return .SocketIpv6
        }else{
            ptr.deallocate()
            return nil
        }
    }
    public var port:UInt16{
        let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: origin.count)
        self.origin.copyBytes(to: ptr, count: origin.count)
        
        let port = tiny_addr_port(addr: ptr, size: self.origin.count)
        ptr.deallocate()
        return port
    }
    public var ip:String {
        let ptr = UnsafeMutablePointer<UInt8>.allocate(capacity: origin.count)
        self.origin.copyBytes(to: ptr, count: origin.count)
        let p = tiny_addr_ip(addr: ptr, size: self.origin.count);
        ptr.deallocate()
        let str = String(cString: p)
        p.deallocate()
        return str
    }
    public init(origin:Data) {
        self.origin = origin
    }
    public init(domain:SocketDomain,ip:String,port:UInt16){
        switch domain {
        case .SocketIpv4:
            let p = tiny_addr_create(domain: Int(AF_INET), addr: ip, port: port)
            let data = Data(bytes: p, count: MemoryLayout<sockaddr_in>.size)
            self.origin = data
            p.deallocate()
        case .SocketIpv6:
            let p = tiny_addr_create(domain: Int(AF_INET6), addr: ip, port: port)
            let data = Data(bytes: p, count: MemoryLayout<sockaddr_in6>.size)
            self.origin = data
            p.deallocate()
            break
        }
    }
}

public func tiny_host(name:String)->SocketHost?{
    guard let end = gethostbyname(name) else { return nil }
    var ptr = end.pointee.h_aliases
    var alias:[String] = []
    while ptr?.pointee != nil {
        alias.append(String(cString: ptr!.pointee!))
        ptr = ptr?.advanced(by: 1)
    }
    
    
    ptr = end.pointee.h_addr_list
    var address:[String] = []
    var addrData:[Data] = []
    let p = UnsafeMutablePointer<Int8>.allocate(capacity: 32)
    while ptr?.pointee != nil {
        inet_ntop(end.pointee.h_addrtype, ptr!.pointee!, p, 32)
        address.append(String(cString: p))
        let data = Data(bytes: ptr!.pointee!, count: Int(end.pointee.h_length))
        addrData.append(data)
        ptr = ptr?.advanced(by: 1)
        bzero(p, 32)
    }
    freehostent(end)
    p.deallocate()
    return SocketHost(name: String(cString: end.pointee.h_name), aliases: alias, length: Int(end.pointee.h_length), addrType: end.pointee.h_addrtype == AF_INET6 ? .SocketIpv6 : .SocketIpv4, ipStrAddr: address, ipAddr: addrData)
}
public func tiny_host()->SocketHost?{
    guard let end = gethostent() else { return nil }
    var ptr = end.pointee.h_aliases
    var alias:[String] = []
    while ptr?.pointee != nil {
        alias.append(String(cString: ptr!.pointee!))
        ptr = ptr?.advanced(by: 1)
    }
    
    
    ptr = end.pointee.h_addr_list
    var address:[String] = []
    var addrData:[Data] = []
    let p = UnsafeMutablePointer<Int8>.allocate(capacity: 32)
    while ptr?.pointee != nil {
        inet_ntop(end.pointee.h_addrtype, ptr!.pointee!, p, 32)
        address.append(String(cString: p))
        let data = Data(bytes: ptr!.pointee!, count: Int(end.pointee.h_length))
        addrData.append(data)
        ptr = ptr?.advanced(by: 1)
        bzero(p, 32)
    }
    freehostent(end)
    p.deallocate()
    return SocketHost(name: String(cString: end.pointee.h_name), aliases: alias, length: Int(end.pointee.h_length), addrType: end.pointee.h_addrtype == AF_INET6 ? .SocketIpv6 : .SocketIpv4, ipStrAddr: address, ipAddr: addrData)
}

public func tiny_pton(ip:String,socket:SocketDomain)->Data {
    switch socket {
    case .SocketIpv4:
        let p = UnsafeMutablePointer<UInt8>.allocate(capacity: MemoryLayout<in_addr>.size)
        inet_pton(AF_INET , ip, p)
        let data = Data(bytes: p, count: MemoryLayout<in_addr>.size)
        p.deallocate()
        return data
    case .SocketIpv6:
        let p = UnsafeMutablePointer<UInt8>.allocate(capacity: MemoryLayout<in6_addr>.size)
        inet_pton(AF_INET6 , ip, p)
        let data = Data(bytes: p, count: MemoryLayout<in6_addr>.size)
        p.deallocate()
        return data
    }
}
public func tiny_ntop(data:Data,socket:SocketDomain)->String {
    let p = UnsafeMutablePointer<UInt8>.allocate(capacity: data.count)
    data.copyBytes(to: p, count: data.count)
    let r = UnsafeMutablePointer<Int8>.allocate(capacity: 32)
    switch socket {
    case .SocketIpv4:
        inet_ntop(AF_INET, p, r, 32)
        let str = String(cString: r)
        p.deallocate()
        r.deallocate()
        return str
    case .SocketIpv6:
        inet_ntop(AF_INET, p, r, 32)
        let str = String(cString: r)
        p.deallocate()
        r.deallocate()
        return str
    }
}
public func tiny_create_addr(domain:SocketDomain,addr:String,port:UInt16)->Data{
    switch domain {
    case .SocketIpv4:
        let addr = tiny_addr_create(domain: Int(AF_INET), addr: addr, port: port)
        let data = Data(bytes: addr, count: MemoryLayout<sockaddr_in>.size)
        addr.deallocate()
        return data
    case .SocketIpv6:
        let addr = tiny_addr_create(domain: Int(AF_INET6), addr: addr, port: port)
        let data = Data(bytes: addr, count: MemoryLayout<sockaddr_in6>.size)
        addr.deallocate()
        return data
    }
}

public typealias SocketSendCallBack = (SocketError?)->Void

File: TinySocket/Classes/socket.c
//
//  socketUtil.c
//  TinySocket
//
//  Created by hao yin on 2021/1/22.
//

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <arpa/inet.h>
#include <sys/types.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <dirent.h>
#include <netdb.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/select.h>
#include <sys/ioctl.h>
#include <ifaddrs.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#define MAXBUFF 4 * 1024

int tiny_tcp(int domain){
    return socket(domain, SOCK_STREAM, IPPROTO_TCP);
}

int tiny_udp(int domain){
    return socket(domain, SOCK_DGRAM, IPPROTO_UDP);
}

int tiny_host_net_family(int family){
    struct ifaddrs * info;
    getifaddrs(&info);
    while (info != NULL) {
        if (info->ifa_addr->sa_family == family){
            return 1;
        }
    }
    return 0;
}
void tiny_send_timeout(int tcp,int seconds){
    setsockopt(tcp,SOL_SOCKET, SO_SNDTIMEO, &seconds, sizeof(int));
}
void tiny_connect_timeout(int tcp,int seconds){
    setsockopt(tcp,SOL_SOCKET, TCP_CONNECTIONTIMEOUT, &seconds, sizeof(int));
}
void tiny_recv_timeout(int tcp,int seconds){
    setsockopt(tcp,SOL_SOCKET, SO_RCVTIMEO, &seconds, sizeof(int));
}
int tiny_tcp_connect(int tcp,int domain,const char *ip,unsigned short port) {
    
    if (domain == AF_INET6){
        struct sockaddr_in6 ips;
        inet_pton(domain, ip, &ips.sin6_addr);
        ips.sin6_port = htons(port);
        ips.sin6_family = AF_INET6;
        return connect(tcp, (struct sockaddr *)&ips, sizeof(ips));
    }else{
        struct sockaddr_in ips;
        inet_pton(domain, ip, &ips.sin_addr);
        ips.sin_port = htons(port);
        ips.sin_family = AF_INET;
        return connect(tcp, (struct sockaddr *)&ips, sizeof(ips));
        
    }
}

size_t tiny_send(int socket,const void * data,size_t size){
    return send(socket, data, size, 0);
}

size_t tiny_send_string(int socket,const char * str){
    return send(socket, str, strlen(str), 0);
}
size_t tiny_recv(int socket,void *data,size_t size){
    
    return recv(socket, data, size, 0);
}

size_t tiny_recv_from(int socket,void *data,size_t size,const void **client,unsigned int * len){
    void * result = malloc(28);
    uint vlen = 28;
    size_t rsize = recvfrom(socket, data, size, 0, result, &vlen);
    
    *client = result;
    
    
    return rsize;
}
size_t tiny_recv_string(int socket,char *str,size_t size){
    
    return recv(socket, str, size, 0);
}

int tiny_close(int socket){
    return close(socket);
}


size_t tiny_send_to(int socket,
                 int domain,
                 const char *ip,
                 unsigned short port,
                 const void * data,
                 size_t size){
    
    if (domain == AF_INET6) {
        struct sockaddr_in6 ipadd;
        bzero(&ipadd, sizeof(ipadd));
        int i = inet_pton(AF_INET6, ip, &ipadd.sin6_addr);
        if(i == 1){
            ipadd.sin6_port = htons(port);
            ipadd.sin6_family = AF_INET6;
            return sendto(socket,data,size, 0, (struct sockaddr *)&ipadd, sizeof(ipadd));
        }
        return 0;
    }else{
        struct sockaddr_in ipadd;
        bzero(&ipadd, sizeof(ipadd));
        int i = inet_pton(AF_INET, ip, &ipadd.sin_addr);
        if(i == 1){
            ipadd.sin_port = htons(port);
            ipadd.sin_family = AF_INET;
            return sendto(socket,data,size, 0, (struct sockaddr *)&ipadd, sizeof(ipadd));
        }
        return  0;
    }
}
size_t tiny_send_string_to(int socket,
                           int domain,
                           const char *ip,
                           unsigned short port,
                           const char * data){
    size_t len = strlen(data);
    return tiny_send_to(socket, domain, ip, port, data, len);
}
int tiny_tcp_accept(int tcp,const void **client,unsigned int * len){
    struct sockaddr * sock = malloc(28);
    *len = 28;
    int socket = accept(tcp, sock, len);
    *client = sock;
    return socket;
}
int tiny_tcp_listen(int tcp,int count) {
    return listen(tcp, count);
}
int tiny_tcp_bind(int tcp,int domain,short port){
    
    if(domain == AF_INET6){
        struct sockaddr_in6 sock;
        bzero(&sock.sin6_addr, sizeof(sock.sin6_addr));
        sock.sin6_family = domain;
        sock.sin6_port = htons(port);
        return bind(tcp, (struct sockaddr *)&sock, sizeof(sock));
    }else{
        struct sockaddr_in sock;
        bzero(&sock.sin_addr, sizeof(sock.sin_addr));
        sock.sin_family = domain;
        sock.sin_port = htons(port);
        return bind(tcp, (struct sockaddr *)&sock, sizeof(sock));
    }
}

int tiny_addr_famaly(const void* addr){
    struct sockaddr * a = (struct sockaddr *)addr;
    return a->sa_family;
}
ushort tiny_addr_port(const void* addr,size_t size){
    if(tiny_addr_famaly(addr) == AF_INET6){
        struct sockaddr_in6 * s = (struct sockaddr_in6 *)addr;
        return ntohs(s->sin6_port);
    }else{
        struct sockaddr_in * s = (struct sockaddr_in *)addr;
        return ntohs(s->sin_port);
    }
}
const char * tiny_addr_ip(const void* addr,size_t size){
    if(tiny_addr_famaly(addr) == AF_INET6){
        struct sockaddr_in6 * s = (struct sockaddr_in6 *)addr;
        char *str = malloc(28);
        bzero(str, 28);
        inet_ntop(AF_INET6, &s->sin6_addr, str, 28);
        return str;
    }else{
        struct sockaddr_in * s = (struct sockaddr_in *)addr;
        char *str = malloc(16);
        bzero(str, 16);
        inet_ntop(AF_INET, &s->sin_addr, str, 16);
        return str;
    }
}

const void *tiny_addr_create(int domain,const char* addr,ushort port){
    if(domain == AF_INET6){
        struct sockaddr_in6 *ip = (struct sockaddr_in6 *)malloc(sizeof(struct sockaddr_in6));
        inet_pton(domain, addr, &ip->sin6_addr);
        ip->sin6_port = htons(port);
        ip->sin6_family = AF_INET6;
        return ip;
    }else{
        struct sockaddr_in *ip = (struct sockaddr_in *)malloc(sizeof(struct sockaddr_in));
        inet_pton(domain, addr, &ip->sin_addr);
        ip->sin_port = htons(port);
        ip->sin_family = AF_INET;
        return ip;
    }
}
